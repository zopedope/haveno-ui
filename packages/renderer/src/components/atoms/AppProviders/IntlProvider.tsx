// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { useEffect, useMemo, useState } from "react";
import { IntlProvider as ReacIntlProvider } from "react-intl";
import type { FC } from "react";
import { LangPack, SupportedLocales } from "@src/constants/lang";
import type { LangKeys } from "@src/constants/lang";

const DEFAULT_LOCALE = SupportedLocales.EN;

export const IntlProvider: FC = ({ children }) => {
  const [messages, setMessages] = useState<{ [key in LangKeys]: string }>();
  const locale = useMemo(
    () =>
      navigator.language in SupportedLocales
        ? navigator.language
        : DEFAULT_LOCALE,
    [navigator.language]
  );

  useEffect(() => {
    setMessages(LangPack[locale] ?? LangPack[DEFAULT_LOCALE]);
  }, [locale]);

  return (
    <ReacIntlProvider locale={locale} messages={messages}>
      {children}
    </ReacIntlProvider>
  );
};
