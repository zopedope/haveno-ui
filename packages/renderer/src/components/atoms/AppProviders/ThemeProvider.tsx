// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { MantineProvider, Global } from "@mantine/core";
import type { FC } from "react";
import { themeOverride, globalStyles } from "@src/theme";

export const ThemeProvider: FC = ({ children }) => {
  return (
    <MantineProvider theme={themeOverride}>
      <Global styles={globalStyles} />
      {children}
    </MantineProvider>
  );
};
