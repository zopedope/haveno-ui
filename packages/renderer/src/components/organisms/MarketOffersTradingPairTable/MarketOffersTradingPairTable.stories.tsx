// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import type { ComponentStory, ComponentMeta } from "@storybook/react";
import { MarketOffersTradingPairTable } from "./MarketOffersTradingPairTable";
import type { TMarketOffersTradingPair } from "./_types";

export default {
  title: "organisms/MarketOffersTradingPairTable",
  component: MarketOffersTradingPairTable,
} as ComponentMeta<typeof MarketOffersTradingPairTable>;

const Template: ComponentStory<typeof MarketOffersTradingPairTable> = (
  args
) => {
  return <MarketOffersTradingPairTable data={args.data} />;
};

export const Default = Template.bind({});

Default.args = {
  data: [
    {
      fromPair: "EUR",
      toPair: "XMR",
      lastPrice: 101.122,
      lastPriceCurrency: "EUR",
      dayChangeRate: 12.12,
      dayChangeVolume: 1222.123,
    },
    {
      fromPair: "EUR",
      toPair: "XMR",
      lastPrice: 101.122,
      lastPriceCurrency: "EUR",
      dayChangeRate: 12.12,
      dayChangeVolume: 1222.123,
    },
    {
      fromPair: "EUR",
      toPair: "XMR",
      lastPrice: 101.122,
      lastPriceCurrency: "EUR",
      dayChangeRate: 12.12,
      dayChangeVolume: 1222.123,
    },
    {
      fromPair: "EUR",
      toPair: "XMR",
      lastPrice: 101.122,
      lastPriceCurrency: "EUR",
      dayChangeRate: 12.12,
      dayChangeVolume: 1222.123,
    },
  ] as Array<TMarketOffersTradingPair>,
};
