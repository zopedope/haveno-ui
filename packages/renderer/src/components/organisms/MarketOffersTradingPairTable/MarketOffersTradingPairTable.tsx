// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { useIntl } from "react-intl";
import { createStyles } from "@mantine/core";
import type { TMarketOffersTradingPair } from "./_types";
import { marketTradingPairTable } from "./_types";
import {
  MarketOfferPair24thChange,
  MarketOfferPairLastPriceCell,
  MarketOfferPair24thChangeVolume,
} from "./MarketOffersTradingPairTableCells";
import { pairColumnAccessor } from "./_utils";
import type { TableProps } from "@molecules/Table";
import { Table } from "@molecules/Table";
import { LangKeys } from "@constants/lang";

export interface MarketOffersTradingPairTableProps extends Partial<TableProps> {
  data: Array<TMarketOffersTradingPair>;
}

export function MarketOffersTradingPairTable({
  data,
  ...rest
}: MarketOffersTradingPairTableProps) {
  const { classes } = useStyles();
  const columns = useMarketTradingPairsColumns();

  return (
    <Table
      {...rest}
      table={marketTradingPairTable}
      columns={columns}
      data={data}
      tableWrap={{
        verticalSpacing: "md",
        highlightOnHover: true,
        className: classes.root,
      }}
    />
  );
}

const useMarketTradingPairsColumns = () => {
  const { formatMessage } = useIntl();

  return [
    marketTradingPairTable.createDataColumn(pairColumnAccessor, {
      id: "pair",
      header: formatMessage({
        id: LangKeys.MarketTradingPairColPair,
        defaultMessage: "Pair",
      }),
      size: 400,
    }),
    marketTradingPairTable.createDataColumn("lastPrice", {
      id: "lastPrice",
      header: formatMessage({
        id: LangKeys.MarketTradingPairColLastPrice,
        defaultMessage: "Last Price",
      }),
      size: 400,
      cell: ({ row }) => <MarketOfferPairLastPriceCell row={row?.original} />,
    }),
    marketTradingPairTable.createDataColumn("dayChangeRate", {
      id: "dayChangeRate",
      header: formatMessage({
        id: LangKeys.MarketTradingPairColDayChange,
        defaultMessage: "24th Change",
      }),
      size: 400,
      cell: () => <MarketOfferPair24thChange />,
      meta: { textAlign: "right" },
    }),
    marketTradingPairTable.createDataColumn("dayChangeVolume", {
      id: "dayChangeVolume",
      header: formatMessage({
        id: LangKeys.MarketTradingPairColDayChangeVolume,
        defaultMessage: "24h Vol",
      }),
      size: 400,
      cell: ({ row }) => (
        <MarketOfferPair24thChangeVolume row={row?.original} />
      ),
      meta: { textAlign: "right" },
    }),
  ];
};

const useStyles = createStyles((theme) => ({
  root: {
    paddingTop: 20,
    paddingBottom: 0,

    thead: {
      tr: {
        th: {
          color: theme.colors.gray[9],
          fontSize: theme.fontSizes.xs,
          paddingTop: 8,
          paddingBottom: 8,
          textTransform: "uppercase",

          "&:first-of-type": {
            paddingLeft: theme.spacing.xl,
          },
          "&:last-of-type": {
            paddingRight: theme.spacing.xl,
          },
        },
      },
    },
    tbody: {
      tr: {
        td: {
          borderBottomColor: "transparent",
          fontSize: theme.spacing.sm * 1.168,
          fontWeight: 600,

          "&:first-of-type": {
            paddingLeft: theme.spacing.xl,
          },
          "&:last-of-type": {
            paddingRight: theme.spacing.xl,
          },
        },
      },
    },
  },
}));
