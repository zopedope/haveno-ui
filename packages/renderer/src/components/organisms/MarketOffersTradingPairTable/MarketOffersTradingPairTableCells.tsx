// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { Box, Group } from "@mantine/core";
import type { TMarketOffersTradingPair } from "./_types";
import { AmountChange } from "@atoms/AmountChange/AmountChange";
import { Currency } from "@atoms/Currency";

export function MarketOfferPairLastPriceCell({
  row,
}: {
  row?: TMarketOffersTradingPair;
}) {
  return (
    <Group spacing="md">
      <Box>{row?.lastPriceCurrency}</Box>
      <Box>
        <Currency value={row?.lastPrice || 0} minimumFractionDigits={0} />
      </Box>
    </Group>
  );
}

export function MarketOfferPair24thChange() {
  return <AmountChange positive={true}>+3,5%</AmountChange>;
}

export function MarketOfferPair24thChangeVolume({
  row,
}: {
  row?: TMarketOffersTradingPair;
}) {
  return (
    <Currency value={row?.dayChangeVolume || 0} minimumFractionDigits={0} />
  );
}
