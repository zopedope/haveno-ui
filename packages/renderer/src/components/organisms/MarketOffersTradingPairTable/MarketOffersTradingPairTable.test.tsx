// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { describe, expect, it } from "vitest";
import { render, screen } from "@testing-library/react";
import type { TMarketOffersTradingPair } from "./_types";
import { MarketOffersTradingPairTable } from "./MarketOffersTradingPairTable";
import { AppProviders } from "@atoms/AppProviders";

describe("molecules::MarketoffersTradingPairTable", () => {
  it("renders without exploding", () => {
    const { asFragment, unmount } = render(
      <AppProviders>
        <MarketOffersTradingPairTable data={data} />
      </AppProviders>
    );
    expect(asFragment()).toMatchSnapshot();
    unmount();
  });

  it("renders all columns", () => {
    const { unmount } = render(
      <AppProviders>
        <MarketOffersTradingPairTable data={data} />
      </AppProviders>
    );
    expect(screen.queryByText("Pair")).toBeInTheDocument();
    expect(screen.queryByText("Last Price")).toBeInTheDocument();
    expect(screen.queryByText("24th Change")).toBeInTheDocument();
    expect(screen.queryByText("24th Vol")).toBeInTheDocument();
    unmount();
  });

  it("renders Pair cells", () => {
    const { unmount } = render(
      <AppProviders>
        <MarketOffersTradingPairTable data={data} />
      </AppProviders>
    );
    expect(screen.queryByText("EUR/XMR")).toBeInTheDocument();
    expect(screen.queryByText("USD/XMR")).toBeInTheDocument();
    unmount();
  });

  it("renders Day Change Volume cells", () => {
    const { unmount } = render(
      <AppProviders>
        <MarketOffersTradingPairTable data={data} />
      </AppProviders>
    );
    expect(screen.queryByText("4,233.123")).toBeInTheDocument();
    expect(screen.queryByText("1,222.123")).toBeInTheDocument();
    unmount();
  });
});

const data = [
  {
    fromPair: "EUR",
    toPair: "XMR",
    lastPrice: 101.12,
    lastPriceCurrency: "EUR",
    dayChangeRate: 12.12,
    dayChangeVolume: 4233.123,
  },
  {
    fromPair: "USD",
    toPair: "XMR",
    lastPrice: 105.12,
    lastPriceCurrency: "EUR",
    dayChangeRate: 83.12,
    dayChangeVolume: 1222.123,
  },
] as Array<TMarketOffersTradingPair>;
