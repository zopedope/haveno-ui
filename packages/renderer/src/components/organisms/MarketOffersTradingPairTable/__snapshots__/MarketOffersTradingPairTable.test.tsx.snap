// Vitest Snapshot v1

exports[`molecules::MarketoffersTradingPairTable > renders without exploding 1`] = `
<DocumentFragment>
  <table
    class="mantine-Table-root __mantine-ref-hover mantine-Table-hover mantine-1u8kzyn"
  >
    <thead>
      <tr>
        <th
          colspan="1"
          style="width: 400px;"
        >
          Pair
        </th>
        <th
          colspan="1"
          style="width: 400px;"
        >
          Last Price
        </th>
        <th
          colspan="1"
          style="width: 400px; text-align: right;"
        >
          24th Change
        </th>
        <th
          colspan="1"
          style="width: 400px; text-align: right;"
        >
          24th Vol
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          style="width: 400px;"
        >
          EUR/XMR
        </td>
        <td
          style="width: 400px;"
        >
          <div
            class="mantine-Group-root mantine-6y1794"
          >
            <div
              class="mantine-Group-child mantine-1oprzqz"
            >
              EUR
            </div>
            <div
              class="mantine-Group-child mantine-1oprzqz"
            >
              101.12
            </div>
          </div>
        </td>
        <td
          style="width: 400px; text-align: right;"
        >
          <div
            class="mantine-ozfbol"
          >
            +3,5%
          </div>
        </td>
        <td
          style="width: 400px; text-align: right;"
        >
          4,233.123
        </td>
      </tr>
      <tr>
        <td
          style="width: 400px;"
        >
          USD/XMR
        </td>
        <td
          style="width: 400px;"
        >
          <div
            class="mantine-Group-root mantine-6y1794"
          >
            <div
              class="mantine-Group-child mantine-1oprzqz"
            >
              EUR
            </div>
            <div
              class="mantine-Group-child mantine-1oprzqz"
            >
              105.12
            </div>
          </div>
        </td>
        <td
          style="width: 400px; text-align: right;"
        >
          <div
            class="mantine-ozfbol"
          >
            +3,5%
          </div>
        </td>
        <td
          style="width: 400px; text-align: right;"
        >
          1,222.123
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
